Feature/Parametric Torus Mesh

otmanesabir opened the pull request:
### General:

I was trying to further test CSG and I thought it was nice to include a torus mesh that we can generate directly from OpenRT. Here are a few screenshots from the tests and demos i tried to run:

### **1) The minimum number of sides we can supply is 3:**
![image](https://user-images.githubusercontent.com/41652846/108889350-923ba680-760c-11eb-8bb2-a7d6cb346241.png)


### **2) Increasing the number of sides:**
![image](https://user-images.githubusercontent.com/41652846/108893909-e8f7af00-7611-11eb-90fb-b98068d8de0f.png)




### **3) Changing the cross-sectional radius to be small:**
![image](https://user-images.githubusercontent.com/41652846/108889452-af707500-760c-11eb-87a1-baffb43b3a6e.png)


### **4) Using Phong shader with smoothing on:**
![image](https://user-images.githubusercontent.com/41652846/108889539-ca42e980-760c-11eb-8424-d8ab1de4dbc2.png)


### **5) Using Eyelight:**
![image](https://user-images.githubusercontent.com/41652846/108889597-dcbd2300-760c-11eb-8e34-d723572368dd.png)


There are also some minor unit tests to check if everything works as in the construction.

Commit:
add torus logic

Commit:
add new class in include

Commit:
solidTorus change assert mistake

Commit:
fix codefactor issues

Commit:
fix conversion issues

Commit:
switch from array to vector

ereator commented on modules/core/SolidTorus.cpp:
```diff
@@ -0,0 +1,53 @@
+#include "SolidTorus.h"
+#include "PrimTriangle.h"
+
+rt::CSolidTorus::CSolidTorus(const rt::ptr_shader_t &pShader, const Vec3f &origin, float r1, float r2, int nSides,
```
Here we can distinguish the torus sides and segments.
If we put a torus at the table and cut it as a cake to pieces - these part are called _segments_
If we take one such piece - it will link like a deformed pipe of cylinder. The number of sides of such cylinder is the number of _sides_. usually `int nSegments = 2 * nSides` I think that the loops for i and j should be adjusted, that one of them is iterated till 2 * nSides...

ereator commented on modules/core/SolidTorus.h:
```diff
@@ -0,0 +1,41 @@
+        * @brief A helper struct to store and access the torus topology when constructing.
+        * @var torus_data:t_vertex
+        */
+        struct torus_data {
```
If this structure is used only in the constructor, perhaps it would be better to move it from the header to the cpp file...

ereator reviewed (commented):
Good job!
Please see just minor comments to the code.

- One more thing, just as info. In 3DsMax it is possible to have r2 > r1. The inner part of the torus then intersects and in my opinion it makes no sense.
- The test renders are cool. Just the eyesight shader looks strange, like there is something wrong with normals. Please check.

otmanesabir committed:
fixed solid logic

otmanesabir committed:
fixed torus unit test

otmanesabir commented:
Thank you for your review prof. @ereator!

* I've changed the number of segments accordingly and also moved the torus_data struct to the implementation file.
* I ran a few tests with textures and found a bag that should be now fixed -> see attached img1. 
* I didn't change the normals because I ran a few more tests and it seemed fine? -> see attached img2.
* I rendered a few more images and played around with the different shaders and lighting modes (all other images) -> very cool 💯 !!

Image 1:
![image](https://user-images.githubusercontent.com/41652846/111140113-3ec0d680-8582-11eb-9d91-1934d3f9356a.png)

Image 2:
![image](https://user-images.githubusercontent.com/41652846/111140836-15ed1100-8583-11eb-8cd9-f808e869d9c6.png)

Other Images:
![image](https://user-images.githubusercontent.com/41652846/111140892-2604f080-8583-11eb-958f-2d33ffa0be98.png)
![image](https://user-images.githubusercontent.com/41652846/111140905-2a310e00-8583-11eb-801d-62d4f9438b6e.png)

otmanesabir committed:
Merge branch 'master' into feature/parametric-torus-mesh

ereator commented on modules/core/SolidTorus.cpp:
```diff
@@ -0,0 +1,62 @@
+                                                                       sin(phi) * (r1 + r2 * cos(theta)),
+                                                                       r2 * sin(theta));
+            torus_vertices[(nSegments * i) + j].t_uv_coord = Vec2f((float)i / (float)nSegments, (float)j / (float)nSides);
+            torus_vertices[(nSegments * i) + j].t_normal = Vec3f(cos(phi) * cos(theta), sin(theta) * cos(phi), sin(theta));
```
I think there might still be the problems with the normals. 
In the CTrinagle the normal depends on the order of the vertices. Please check.

otmanesabir replied:
I just took a look at the new changes and it makes sense. Thank you for refactoring the code as well! I have no open comments from my side

ereator reviewed (changes requested):
Please my comments in Slack about the artifacts with the normals

ereator committed:
Fix normal issue

ereator committed:
Update SolidTorus.cpp

ereator commented:
I changed the normal estimation from 
`Vec3f(cos(phi) * cos(theta), sin(theta) * cos(phi), sin(theta));` 
to
`Vec3f(cosf(phi) * cosf(theta), sinf(phi) * cosf(theta), sinf(theta));`
I believe it was just a typo.
